src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -311,125 +309,81 @@ void FileBasedWal::prepareNewFile(LogID startLogId) {
-                return false;
-            }
-            size -= res;
-            start += res;
-        } while (size > 0);
-        return true;
-    };
-    if (!cord.applyTo(cb)) {
-        LOG(FATAL) << "Failed to flush the wal file";
-    } else {
-        // Succeeded writing all buffered content, adjust the file size
-        currInfo_->setSize(currInfo_->size() + cord.size());
-        currInfo_->setLastId(lastId);
-        currInfo_->setLastTerm(lastTerm);
+TermID FileBasedWal::readTermId(const char* path, LogID logId) {
+    int32_t fd = open(path, O_EXCL | O_RDONLY | O_CLOEXEC | O_LARGEFILE, 0644);
+    if (fd < 0) {
+        LOG(FATAL) << "Failed to open file \"" << path
+                   << "\" (errno: " << errno << "): "
+                   << strerror(errno);
     }
-}
 
+    size_t pos = 0;
+    while (true) {
+        // Read the log Id
+        LogID id = 0;
+        if (pread(fd, &id, sizeof(LogID), pos) != sizeof(LogID)) {
+            LOG(ERROR) << "Failed to read the log id (errno "
+                       << errno << "): " << strerror(errno);
+            close(fd);
+            return 0;
+        }
 
-void FileBasedWal::flushBuffer(BufferPtr buffer) {
-    std::lock_guard<std::mutex> flushGuard(flushMutex_);
-    if (!buffer || buffer->empty() || buffer->invalid()) {
-        return;
-    }
+        // Read the term Id
+        TermID term = 0;
+        if (pread(fd, &term, sizeof(TermID), pos + sizeof(LogID)) != sizeof(TermID)) {
+            LOG(ERROR) << "Failed to read the term id (errno "
+                       << errno << "): " << strerror(errno);
+            close(fd);
+            return 0;
+        }
 
-    // Rollover if required
-    if (buffer->needToRollover()) {
-        closeCurrFile();
-    }
+        if (id == logId) {
+            // Found
+            return term;
+        }
 
-    Cord cord;
-    LogID firstIdInCord = buffer->firstLogId();
-    auto accessFn = [&cord, &firstIdInCord, this] (
-            LogID id,
-            TermID term,
-            ClusterID cluster,
-            const std::string& log) {
-        cord << id << term << int32_t(log.size()) << cluster;
-        cord.write(log.data(), log.size());
-        cord << int32_t(log.size());
-
-        size_t currSize = currFd_ >= 0 ? currInfo_->size() : 0;
-        if (currSize + cord.size() > maxFileSize_) {
-            dumpCord(cord, firstIdInCord, id, term);
-            // Reset the cord
-            cord.clear();
-            firstIdInCord = id + 1;
-
-            // Need to close the current file and create a new file
-            closeCurrFile();
-        }
-    };
-
-    // Dump the buffer to file
-    auto lastLog = buffer->accessAllLogs(accessFn);
-
-    // Dump the rest if any
-    if (!cord.empty()) {
-        dumpCord(cord, firstIdInCord, lastLog.first, lastLog.second);
-    }
+        // Read the message length
+        int32_t len;
+        if (pread(fd, &len, sizeof(int32_t), pos + sizeof(LogID) + sizeof(TermID))
```

Comment:
maybe we should read the termId from end to begin, because there are duplicate log id in file when we call "FileBasedWal::appendLogs" failed.

Reply:
There will be no dup log id in the same file. When rolling back, a new WAL file will be created to overwrite the log ids in the previous file